### ISSUE TITLE
[SPIKE] Discuss Checkout Composable

### ISSUE THREAD

**Issue by filrak** [DivanteLtd/vue-storefront#4576]
## What is the motivation for adding / enhancing this feature?

Currently there is no standardized way to create checkout composable and that's the last missing piece of them. Even though for most of the platforms integrated with VSF Next we are using external checkout as a recommended way there are platforms like commercetools that are purely API-driven and doesn't have this option

Most likely we will need to make this in the same way we made `useuser` composable - split it by smaller entities correlating to specific feature/subpage,

In that case we will (probably) need an interface for
- `useCheckout` for placing orders and order review
- `useCheckoutPersonalDetails`
- `useCheckoutShippingDetails`
- `useCheckoutPayment`
Above 3 are connected to useCheckout

## What are the acceptance criteria 

- [ ] We decided what composables we need for checkout
- [x] We decided on interfaces for those

## TODO After finishing

- Create issue for each composable

**Comment by andrzejewsky** [comment]
First of all, we have to distinguish two things: checkout itself and payment. The checkout is actually a part of eCommerce platform, without that you can't make an order. Is there any platform that doesn't have a checkout? The only thing we delegate externally is a payment system or more (eg. delivery).

Let's start with defining composables and challenge them. I propose the following ones:
- `useCheckoutCustomer` - for defining personal details of user (person who makes an order)
- `useCheckoutShipping` - for defining shipping details of user (where you want to send the order to), and shipping methods
- `useCheckoutPayment` - for defining billing details (data of the user who pays for this), and handling payment process
- `useCheckoutOrder` - for handling an order

Now, below a few potential issues/concerns that needs to be solved:
- all of these composables needs to be connected together with the shared state.
- if you use eg.  `useCheckoutPayment` composable most likely you need the data from the previous ones, how to load that? How to prevent over-fetching?
- billing and shipping are very similar and can be combined into the one composable eg. `useCheckoutAddress`, they are just taking data from the form and saving it to the API, perhaps it will be something more complex further?
- payment is the most complicated part, most of the payment systems provides also special secured UI that uses iframes to load inputs, so you can skip that part - what about this? Is it mean that along with composable we need to provide a component as well? Or we'll provide just a composable, but the component is the developer's responsibility? - think this is the most reasonable
- payment hooks - some of the payment systems require backend (eg. braintree) eg. to validate something, do we provide this as well? nuxt (at least as basic one)?
- confirmations - sometimes an external payment system redirects us somewhere to our app, with some kind of confirmation eg. when payment succeeded. I think with payment integration we should provide also an example of error/success pages that can be replaced by the developer.
-  we have one composable for payment and billing details, of course it makes sense as it's a related thing,  but what if you need to save billing address to the eCommerce platform but payment is managed by something external (that's a real example in CT)? 

To sum up. Checkout is essentially the most difficult part to unify. It has a huge variety of possible implementations. I'm even thinking about preparing totally custom composables per eCommerce integration with the implementation of its own checkout and exactly the same with payment systems.

**Comment by filrak** [comment]
@andrzejewsky this is looking very good - I see that similarly.

For the shared state - I think there is nothing wrong with sharing the state between checkout compostable similarly to what we have with `useUser`. In the end they're just kind of "subcomposables" and the connection is implict.
Also some thoughts about your concerns:


> billing and shipping are very similar and can be combined into the one composable eg. useCheckoutAddress, they are just taking data from the form and saving it to the API, perhaps it will be something more complex further?

We could do that. The only thing here that matters to me is extensibility. We have to build it in a way that allows to easily integrate any 3rd-party shipping provider.

> payment is the most complicated part, most of the payment systems provides also special secured UI that uses iframes to load inputs, so you can skip that part - what about this? Is it mean that along with composable we need to provide a component as well? Or we'll provide just a composable, but the component is the developer's responsibility? - think this is the most reasonable
payment hooks - some of the payment systems require backend (eg. braintree) eg. to validate something, do we provide this as well? nuxt (at least as basic one)?
confirmations - sometimes an external payment system redirects us somewhere to our app, with some kind of confirmation eg. when payment succeeded. I think with payment integration we should provide also an example of error/success pages that can be replaced by the developer.

Yes, I thought about ti this way: each payment integration needs to fulfill some specific contract (let it be a typescript interface) so it can be passed to the setup config and extend proper parts of the app (hooks as you called them)

Apart from that there should always be a component that will have to be used within VSF App in a checkout that has to be added manually (in the future we can probably add it automatically as well)

> we have one composable for payment and billing details, of course it makes sense as it's a related thing, but what if you need to save billing address to the eCommerce platform but payment is managed by something external (that's a real example in CT)?

I don't see how having both as one composable will prevent what you've written about

**Comment by filrak** [comment]
# DRAFT, DISCUSSION WILL BE CONTINUED

Discussion outcome:

0. Use builder pattern to dynamically generate checkout forms (Jacob Schwartz talk from VueJS Amsterdam 2019)
1. `useCheckoutX`
- default data model, configure data model
- getting data from eCommerce (shipping methods, payment methods)

```js
import { configureFormDataModel, formDataModels } from '@vue-storefront/commercetools`

configureFormDataModel(formDataModels.checkout.customer, async currentDataModel => {
  ...currentDataModel,
  firstName: { type: 'input', validation: ['required'] },
  country: { type: 'select', options: currentDataModel.countries.options }
})
```
What if we need to load some data lazily?
```js
configureFormDataModel(formDataModels.checkout.customer, async currentDataModel => {
  ...currentDataModel,
  country: { type: 'select', options: currentDataModel.countries.options, async onChange: (value) => {
  currentDataModel.regions = getRegionsFromApi(value.id)
})

```
### Payment Plugins

```js
// you specify which type of plugin you're registering
import { addPlugin } from '@vue-storefront/commercetools`
import { PaymentPlugin, PaymentComponent } from 'paypal-extension'


addPaymentMethod(PaymentPlugin)

export PaymentPlugin = {
  addPaymentMethod() {
   return { ... }
 }
}
```

**Comment by andrzejewsky** [comment]
Discuss notes (interfaces draft):

```ts


interface FormRow  {}

interface AgnosticForm {
  [field: string]: FormRow
}

interface UseCheckoutCustomer <CUSTOMER, LOAD_PARAMS, SAVE_PARAMS> {
  async loadCheckoutCustomer: (params: LOAD_PARAMS) => void;
  async saveCheckoutCustomer: (params: SAVE_PARAMS) => void;
  checkoutCustomer: ComputedProperty<CUSTOMER>;
  form: AgnosticForm;
  loading: boolean;
  error: any;
}

interface UseCheckoutShipping <SHIPPING, LOAD_PARAMS, SAVE_PARAMS> {
  async loadCheckoutShipping: (params: LOAD_PARAMS) => void;
  async saveCheckoutShipping: (params: SAVE_PARAMS) => void;
  checkoutShipping: ComputedProperty<SHIPPING>;
  form: AgnosticForm;
  loading: boolean;
  error: any;
}

interface UseCheckoutPayment <PAYMENT, LOAD_PARAMS, SAVE_PARAMS> {
  async loadCheckoutPayment: (params: LOAD_PARAMS) => void;
  async saveCheckoutPayment: (params: SAVE_PARAMS) => void;
  checkoutPayment: ComputedProperty<PAYMENT>;
  form: AgnosticForm;
  loading: boolean;
  error: any;
}

interface UseCheckoutOrder <ORDER, LOAD_PARAMS, SAVE_PARAMS> {
  async loadCheckoutOrder: (params: LOAD_PARAMS) => void;
  async saveCheckoutOrder: (params: SAVE_PARAMS) => void;
  checkoutOrder: ComputedProperty<ORDER>;
  form: AgnosticForm;
  loading: boolean;
  error: any;
}


const useCheckoutCustomer = (): UseCheckoutCustomer => {}
const useCheckoutShipping = (): UseCheckoutShipping => {}
const useCheckoutPayment = (): UseCheckoutPayment => {}
const useCheckoutOrder = (): UseCheckoutOrder => {}
```

**Comment by filrak** [comment]
Regarding External Payments - we will most likely pass an object with a unified payment interface rather than a function, apart from that I think we are close to writing actual issues

**Comment by filrak** [comment]
The interface will be implemented as described, the extendibility mechanism discussions will be continued in the next issue
